File: 213-house-robber-ii/213-house-robber-ii.cpp
class Solution {
    int maxProfit(vector<int>& nums, int start, int end) {
        int pick, nonPick;
        int prev = 0, prev2 = 0;

        for(int i = start; i < end; i++) {
            pick = nums[i] + prev2;
            nonPick = prev;
            
            prev2 = prev;
            prev = max(pick, nonPick);
        }
        
        return prev;
    }
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 1) return nums[0];
        
        
        int profitWithFirst = maxProfit(nums, 0, nums.size()-1);
        int profitWithLast = maxProfit(nums, 1, nums.size());
        return max(profitWithFirst, profitWithLast);
    }
};

File: 213-house-robber-ii/213-house-robber-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findProf = function(nums, st, n) {
    let nex = 0, nex2 = 0
    
    for(let i = n-1; i >= st; i--) {
        let pick = nums[i] + nex2
        let notPick = nex
        nex2 = nex
        nex = (pick < notPick) ? notPick : pick
    }
    
    return nex
};

var rob = (nums) => {
    const n = nums.length;
    if(n <= 2) return Math.max(...nums)
    
    let maxProf = Math.max(findProf(nums, 0, n-1), findProf(nums, 1, n))
    
    return maxProf;
}